### ISSUE TITLE
AudioFile Read threads not being released on Android

### ISSUE THREAD

**Issue by lawrencedine** [superpoweredSDK/Low-Latency-Android-iOS-Linux-Windows-tvOS-macOS-Interactive-Audio-Platform#538]
Hello,

We're seeing crashes in our app due to running in to the file descriptor limit on Android, on further investigation it seems that for each time we call `open` on our `SuperpoweredAdvanceAudioPlayer` instances we end up with an `AudioFile Read` thread that is never disposed visible using `adb shell ps -t | grep <pid>`. We're using a version of the SuperpoweredSDK downloaded on the 8th of October.

Our app uses multiple `SuperpoweredAdvancedAudioPlayer` instances to play back an arbitrary number of music parts (files) to create a single track. We are using a single `SuperpoweredAndroidAudioIO` instance and have just updated to re-use `SuperpoweredAdvancedAudioPlayer` instances but this issue still persists. We are preventing loading of another track until the `SuperpoweredAdvancedAudioPlayerCallback` reports Success for all players.

What you see in the app is when a user uses 'next track' or 'previous track' enough times the app will crash due to various file descriptor limit issues (database access/preferences access etc).

Commenting out the call to `open` and executing our callback for all players loaded immediately prevents the crash and the thread count remains stable.

TL;DR: It seems like AudioFile Read threads are not being disposed on Android.
	
Thanks,
Lawrence

**Comment by gaborszanto** [comment]
There should always be at least one "AudioFile Read" thread per each player (after any open() call). You get much more than that?

**Comment by lawrencedine** [comment]
It seems like I'm getting double the number of AudioFile Read threads than I would expect, so for my 2 test tracks here I'm getting 22 new threads for a track with 11 parts (11 calls to `open()`) and I'm getting 30 more threads for a track with 15 parts.

**Comment by lawrencedine** [comment]
I tried to reproduce this issue using the SuperpoweredPlayer example but I was unable to recreate the behaviour I'm seeing in our actual app where the threads don't get disposed, they are definitely being disposed in my modified example. However, I am seeing 2 `AudioFile Read` threads per `open()` call.

I can see some clear differences between this example and the one our app was originally based on, most significantly I can see this one doesn't define a C++ class called `SuperpoweredBridge` but defines all methods and variables as static, I will try refactoring our app to this approach and see if that resolves the issue.

**Comment by gaborszanto** [comment]
You may have two AudioFile Read threads around an open() call, because one corresponds to the previous file and waiting to be destructed.
Perhaps you recreate your instance of "SuperpoweredBridge" every time, or instantiate your native library more than once.

**Comment by lawrencedine** [comment]
Ok I think I've figured this issue out. It wasn't the difference between having a SuperpoweredBridge class or all static functions but that did lead me to enough refactoring of code that I found the actual solution.

So if anyone else has this issue here is what I had to adjust to fix it:

1. Ensure re-use of `SuperpoweredAdvancedAudioPlayer`, recreating those each time new tracks were loaded was definitely incorrect.
2. We are storing our (arbitrary number of) `SuperpoweredAdvancedAudioPlayer` instances in a `std::vector`, instead of using array index access `[]` to store/delete instances using `push_back` and `pop_back` prevents this issue occurring
3. In an attempt to fix a previous bug (likely caused by recreating player instances) code had been added to return false immediately from the `audioProcessing` callback if we weren't playing or audio wasn't loaded, this was causing `process()` on the players to not get called which I believe was causing issues
4. In addition to this in an attempt to fix another past bug the `SuperpoweredAndroidAudioIO` was being `stop()` before and `start()` after the reloading of parts, removing this was necessary.

Now I only see `AudioFile Read` threads equal to 2 times the current number of calls to `open()` for the playing track, the threads from previous tracks seem to be correctly disposed of now.
